### ISSUE TITLE
[unstable option] `imports_granularity`

### ISSUE THREAD

**Issue by 9999years** [rust-lang/rustfmt#4991]
Tracking issue for unstable option: `imports_granularity`.

**Comment by Tamschi** [comment]
When using

```toml
unstable_features = true
imports_granularity = "Crate"
```

, `cargo +nightly fmt` will currently remove comments like this:

```rust
use triomphe::{
	Arc,
	// This crate walks along the graph a whole lot and normally doesn't do much refcounting,
	// so the offset version is likely a bit better here.
	//
	// This is unbenched though. Someone might want to check.
	OffsetArc,
};
```

```
❯ cargo +nightly fmt --version
rustfmt 1.4.37-nightly (a8f2463c 2021-10-09)
```

(I wasn't sure whether this deserves its own issue.)

**Comment by calebcartwright** [comment]
The dropping of comments with the various options that manipulate imports is a known issue, and one of the hurdles that will have to be overcome before the options, including this one, can be stabilized.

**Comment by Xuanwo** [comment]
Hi, is there any progress to make this option stable?

**Comment by calebcartwright** [comment]
> Hi, is there any progress to make this option stable?

The absence of updates should be explicitly interpreted as meaning that there are no updates. The process and requirements for stabilizing an option are both linked, and enumerated with checkboxes, in the issue description and will be updated as-and-when there are updates to share

I appreciate the interest folks have in being able to use an option on stable. However, this remains a relatively new option with many well known bugs, and as stated above, is not going to be eligible for stabilization any time soon.

**Comment by Nutomic** [comment]
For what its worth, we have been using `imports_granularity="Crate"` in [Lemmy](https://github.com/LemmyNet/lemmy/blob/main/.rustfmt.toml) for over a year without any problems. Though we dont put any comments around imports.

**Comment by davidlattimore** [comment]
I sent a PR to make rustfmt preserve comments during flattening when they're on the line before a leaf node, or on the same line as a leaf node. e.g.

```rust
use a::{
   // Before a::b
   b,
   c, // After a::c
};
```

If import_granularity=Item, then we'll get:

```rust
// Before a::b
use a::b;
use a::c; // After a::c
```

What I'm less sure about, is what to do when flattening and there's a comment before a non-leaf node. e.g.

```rust
use a::{
   // Before a::b
   b::{b1, b2},
   c,
};
```

One easy option is to not flatten non-leaf nodes that have comments, then we'd get:

```rust
// Before a::b
use a::b::{b1, b2};
use a::c;
```

That would mean we couldn't fully convert granularity to `Item` in such cases.

Another option would be to attach the comment to the first leaf contained within. Then we'd get:

```rust
// Before a::b
use a::b::b1;
use a::b::b2;
use a::c;
```

Although if there was already a comment attached to `b1`, then I guess we'd need to merge the comments, which could get tricky. Thoughts?

**Comment by Xuanwo** [comment]
> Another option would be to attach the comment to the first leaf contained within.

Seems reasonable to me.

> Although if there was already a comment attached to `b1`,

Not bad to me, though.

Converting from

```rust
use a::{
   // Before a::b
   b::{
        // Before a::b::b1
        b1, 
        b2
   },
   c,
};
```

to

```rust
// Before a::b
// Before a::b::b1
use a::b::b1;
use a::b::b2;
use a::c;
```

**Comment by repi** [comment]
Did a reformatting of a fairly large Rust code base (~200k LoC, 767 source files, 60 crates) with `imports_granularity=item` to reduce git merge conflicts on import changes, and happy to report that with latest rustfmt (used https://github.com/rust-lang/rustfmt/commit/a37d3ab0e1c7c05f1a6410fb7ddf5539f0d030f8) that worked without any problems whatsoever. 🎉

Looking for to latest rustfmt here (with the https://github.com/rust-lang/rustfmt/issues/5030 fix) getting deployed out to Rust Nightly in rustup so it can be used more easily, and (hopefully) soon stabilization of this great option! Thanks to everyone that has contributed to this feature/option!

**Comment by calebcartwright** [comment]
I'll also add that I do view folks sharing their experience to be helpful feedback for us, e.g. https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1046885320 and https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1113996096

That level of direct feedback absolutely helps feed into that 3rd gate of "well tested, optimally in real world usage", and is appreciated

**Comment by calebcartwright** [comment]
Big thanks again to @davidlattimore for addressing the comment issue, and I believe this has us on a good track towards stabilization. I believe there's really only two things left we need to consider before proceeding:

* do we have the ability to stabilize the option while leaving certain variants unstable (also so we can add new unstable variants in the future)
* do any of the current variants have any reported bugs

The latter is something anyone could help with and which would be most appreciated. The former is something we need to look at anyway as that's a more general need than any one config option (though anyone interested in helping with that should certainly feel free)

**Comment by tommilligan** [comment]
> I'll also add that I do view folks sharing their experience to be helpful feedback for us, e.g. [#4991 (comment)](https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1046885320) and [#4991 (comment)](https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1113996096)

I've been using both `imports_granularity` and `group_imports` at work for the last few months, and have had no issues apart from the aforementioned comments issue (now fixed). For reference our Rust codebase is ~79k LoC, ~300 files, ~50 crates.

> - do any of the current variants have any reported bugs

I've searched the current open issues referencing `imports_granularity`, and I believe we can consider this box checked. I found:

- #5360 (new feature request, orthogonal to the current feature/current implementation is forwards compatible)
- #5321 (unrelated)
- #4746 (a bug that is general with rustfmt, not imports_granularity specifically)

**Comment by tommilligan** [comment]
> - do we have the ability to stabilize the option while leaving certain variants unstable (also so we can add new unstable variants in the future)

@calebcartwright If I understand this correctly, what you're requesting is that after stabilisation, it would be possible to:

- Pass an option such as `imports_granularity = "Crate"`, which is considered stable and will run on rustfmt stable
- Pass an option such as `imports_granularity = "NewVariant"`, which is considered unstable, and will not run on rustfmt stable

This behaviour should extend outwards to the publically exported API of the crate, such that adding a new variant is not a breaking change. However, I see that the current config option enums are unreachable from the crate root, so maybe this isn't a concern?

https://github.com/rust-lang/rustfmt/blob/5fa2727ddeef534a7cd437f9e288c221a2cf0b6a/src/config/mod.rs#L15

I would be happy to take a first pass at this, I have some bandwidth today

**Comment by calebcartwright** [comment]
@tommilligan - thank you!

> I've searched the current open issues referencing imports_granularity, and I believe we can consider this box checked. I found:

#5360 would potentially be impactful, depending on if, and how, we decide to handle it. It wouldn't necessarily need to be implemented in order to be able to stabilize this option (fully or partially, for some variants), but would at least need to rule out parts of that proposal in order to proceed with stabilization.

>  what you're requesting is that after stabilisation, it would be possible to:

Correct. I want to make sure that we have the ability to make certain variants available on stable even if other variants aren't quite ready, and I also want us to be able to add new variants in the future which would inherently be unstable when first introduced.

> This behaviour should extend outwards to the publically exported API of the crate, such that adding a new variant is not a breaking change. However, I see that the current config option enums are unreachable from the crate root, so maybe this isn't a concern?

I don't think so, or at least I believe it's something we can handle internally, even if it requires something new like a post-processing/pre-formatting validity check of the config.

> I would be happy to take a first pass at this, I have some bandwidth today

:heart: :rocket:

**Comment by tommilligan** [comment]
> do we have the ability to stabilize the option while leaving certain variants unstable (also so we can add new unstable variants in the future)

I have opened a tracking issue at #5378 and a first implementation pass at #5379

**Comment by calebcartwright** [comment]
Also note that #4725 would be a blocker for stabilizing the `Item` variant (it's closed, but note that it has the backport pending label which means the fix hasn't actually landed on the trunk branch yet)

**Comment by davidlattimore** [comment]
Thanks @calebcartwright - I created a [PR](https://github.com/rust-lang/rustfmt/pull/5380) to backport it.

**Comment by utkarshgupta137** [comment]
> @utkarshgupta137 - it's not clear to me precisely what you are proposing, so perhaps some concrete examples would help visualize things.
> 
> However, it seems as if you're saying that the single, global `imports_granularity` option should actually support divergent granularity levels for different import groups (with the specific group example given as core/std/alloc)?

Exactly.

Current behaviour with `imports_granularity` = `Crate`:
```
use std::{
    fmt,
    io::{self, Write},
    net::{self, TcpStream, ToSocketAddrs},
    ops::DerefMut,
    path::PathBuf,
    str::{from_utf8, FromStr},
    time::Duration,
};

use futures_util::{
    future::{Future, FutureExt},
    ready,
    sink::Sink,
    stream::{self, Stream, StreamExt, TryStreamExt as _},
};
```

Desired behaviour:
```
use std::fmt;
use std::io::{self, Write};
use std::net::{self, TcpStream, ToSocketAddrs};
use std::ops::DerefMut;
use std::path::PathBuf;
use std::str::{from_utf8, FromStr};
use std::time::Duration;

use futures_util::{
    future::{Future, FutureExt},
    ready,
    sink::Sink,
    stream::{self, Stream, StreamExt, TryStreamExt as _},
};
```

**Comment by calebcartwright** [comment]
Thank you for clarifying.

I'll wait to let the masses weigh in, but speaking purely personally, will note that sounds a bit too bespoke and has too much friction with the consistency goals imo.

Though technically possible, I'm not sure how practical it would be (especially since it would open a fairly large door), and it would certainly derail the current stabilization progress.

**Comment by faern** [comment]
I agree with @calebcartwright. Sounds too bespoke and edge casey for me also. What about other crates with a huge API surface in very many unrelated areas? Such as `async-std` for example. Should rustfmt then allow specifying a list of crates to except from the import merge rule?

**Comment by utkarshgupta137** [comment]
I too would prefer that this option gets merged first & then we can consider this later on.

**Comment by calebcartwright** [comment]
> this option gets merged first & then we can consider this later on.

The challenge is that I'm not sure I see a great a way in which we could stabilize this option and then introduce the behavior you're referring to. The only way I can think of that wouldn't necessitate breaking/changing a stabilized option (not permitted) would be to have a rather intense new variant for `imports_granularity` with a very long name that provided what you're asking for.

If that's a solution/approach you could live with in a hypothetical future where the multiple granularities model would be supported then we continue proceeding with stabilization. However, if not and you/others will still want divergent granularities for different groups then a number of things, including this option, would need to be reconsidered

**Comment by tommilligan** [comment]
> If that's a solution/approach you could live with in a hypothetical future where the multiple granularities model would be supported then we can continue proceeding with stabilization.

I can see a path forward where the existing option is considered the `default` formatting style for all imports. A new option can then be added and stabilised which takes a mapping of crate name -> granularity level, overriding the default value.

However, I agree that this is a very complex case and I'm not sure what value it brings to the user. I'm not personally interested it, though I am very interested in having `imports_granularity` stabilised - and I think many users would say the same. I would be keen to press forward with stabilisation, rather than scope creeping the original feature.

**Comment by calebcartwright** [comment]
> I can see a path forward where the existing option is considered the default formatting style for all imports. A new option can then be added and stabilised which takes a mapping of crate name -> granularity level, overriding the default value.

Agree with the thrust of your comment, but this is precisely the type of thing we'd want to avoid as it just leads to a more convoluted config surface that's increasingly difficult to consume and explain. This is even more true for imports where there's already a disproportionately large number of configuration options that have to be considered against each other

**Comment by utkarshgupta137** [comment]
Would it be acceptable to add an attribute, similar to `rustfmt_skip` later on?

Now that I think about it, setting this attribute in each file instead of globally for each crate would make it much more useful.

**Comment by calebcartwright** [comment]
> Would it be acceptable to add an attribute, similar to rustfmt_skip later on?

I understand and appreciate you're just trying to brainstorm @utkarshgupta137, but have to stress this is an absolutely massive creep of scope and is completely unrelated to the stabilization of a configuration option.

I don't want this stabilization tracking issue to go off the rails on an orthogonal topic, but will say that while I understand why some users may want to have attribute-based, divergent formatting within their codebase, that's highly unlikely to ever come to fruition both because of running starkly counter to the consistency goals as well as technical reasons.

n.b. this has been previously requested and discussed elsewhere, so we're not going to rehash any aspect of that discussion/debate here.

<hr />

> there should be an option or some mechanism to skip this for core/std/alloc, at least in the crate mode.

Back to your original ask :point_up:, Thanks for sharing your suggestion, but the overall consensus appears to be against rethinking this option to support your ask, so I'm going to go ahead and say that he answer to your ask is a "no". We're not going to pivot and we'll continue with imports_granularity as-is.

Perhaps your request can be revisited down the road, but will have to be considered against the config surface in it's future form.

**Comment by tommilligan** [comment]
So, to again summarize progress towards stabilization:

---

> Also note that https://github.com/rust-lang/rustfmt/issues/4725 would be a blocker for stabilizing the Item variant

As above, backported in #5380, so I consider this unblocked.

> I want to make sure that we have the ability to make certain variants available on stable even if other variants aren't quite ready, and I also want us to be able to add new variants in the future which would inherently be unstable when first introduced.

This is now implemented in #5379.

Are there any variants of `imports_granularity` that need to be annotated as unstable, to allow us to proceed with stabilisation of the option as a whole? IIRC in the past there has been some discussion about the `Item` variant being unstable, but as above I think these are now fixed and backported. I would consider all of the variants stable at present.

> #5360 would potentially be impactful, depending on if, and how, we decide to handle it. It wouldn't necessarily need to be implemented in order to be able to stabilize this option (fully or partially, for some variants), but would at least need to rule out parts of that proposal in order to proceed with stabilization.

There has now been suitable discussion on the PR, which resulted in [(ref)](https://github.com/rust-lang/rustfmt/issues/5360#issuecomment-1152817245):
```
What I might be willing to consider is a new variant for imports_granularity that's something
akin to OneModule (or anything better named) that would provide your requested output.
```

This sounds future-compatible with the unstable variants work that is now completed, so I don't consider this blocking. If more work is required it can be behind an unstable variant.

> https://github.com/rust-lang/rustfmt/issues/5410

There was some discussion about whether this impacts `imports_granularity`; in summary, no it does not. The solution could be "to introduce yet another imports related config option", if desired (so not relevant to stabilisation here).

---

From the tracking list above:

> The option is well tested, both in unit tests and, optimally, in real usage.

There have been several real-world codebases formatted with this option, with no current issues noted:

- https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1046885320
- https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1113996096
- https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-1152157410

> There is no open bug about the option that prevents its use.

I noted #5269 which potentially impacts all import options (sort order). However, there does not seem to be pressure or activity towards resolving it at present.

I think we could either:
- Continue, considering import sort order orthogonal to this feature.
- Block on fixing the issue as a whole.
- Block on fixing #5296 for specific `imports_granularity` functionality. This would ensure the behaviour of `imports_granularity` doesn't change in the future; however, it may make sort order inconsistent with other current rustfmt features.

---

Apologies for the wall of text, but I think that's everything to consider in one place. IMO the only potential blocker remaining is to consider the impact of #5269, which is a far-reaching issue (not specific to this feature)

**Comment by calebcartwright** [comment]
Thank you for the summary @tommilligan and generally agreed.

I want to do a final pass through things, but just so folks know where my head is at, I'd anticipate all variants except `One` being ready for stabilization.

To be clear, it's not that I necessarily think that variant is explicitly _not_ ready, but simply that I'm not yet convinced that it is ready, especially compared to the others, and we err on the side of caution when it comes to stabilization.

**Comment by laralove143** [comment]
this suddenly broke on me idk why, i ran `cargo fmt` and it complained it cant set the option on stable so its seeing the config

/Users/lara/Library/Preferences/rustfmt/rustfmt.toml
```toml
unstable_features = true
imports_granularity = "Crate"
```

right now it's just not doing anything, if the code is
```rust
use crate::Foo;
use crate::Bar;
```
and i run `cargo +nightly fmt` the code stays the same 

i've done `rustup update` and even reinstalled the components, also the `group_imports` doesnt seem to be working either, if there's no empty lines it just treats them all the same (sorts std somewhere in the middle)

idk what broke it and it might be something i overlooked but help please

**Comment by ytmimi** [comment]
@laralove143  I'd suggest opening up a new issue to discuss the problem that you're facing. You should probably include all these details in the issue as well as what version of rustfmt you're using. `cargo +nightly fmt --version` should help you figure out the version.

**Comment by ytmimi** [comment]
@calebcartwright I'm working to backport #3999 in #5290

**Comment by tgross35** [comment]
Would an option like `FlatCrate` be possible, or would that be a separate feature flag?

For example, the `Crate` option has this as an example:

```rust
use foo::{
    a, b,
    b::{f, g},
    c,
    d::e,
};
use qux::{h, i};
```

But that's really bulky.  Id be happier with something like this, that wraps at the line length:

```rust
use foo::{a, b, b::{f, g}, c, d::e,};
use qux::{h, i};
```

Obviously the difference is pretty dramatic in this example with single-letter modules, but it makes a difference in real-world use too.

**Comment by dcormier** [comment]
@tgross35, I believe you're looking for [the `imports_layout` setting](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#imports_layout).

**Comment by tgross35** [comment]
Hm, you're right, but it doesn't seem like the 'Crate" imports_granularity setting is quite coherent. I wrote it up on the `imports_layout` issue https://github.com/rust-lang/rustfmt/issues/3361#issuecomment-1494727534

**Comment by gilescope** [comment]
I don't think the default value is correct - it takes up a lot of space. I expect something like imports_granularity = "Crate". At the moment cargo fmt by default can have pages of use statements with one import on each line. That seems an odd default... (totally happy with every other cargo defalut fmt setting, but this one I just seem to have to set)

**Comment by popzxc** [comment]
Hey! I'm happy to see the rustfmt project active once again.

Can we resume the discussion on whether it's possible to stabilize this feature in some form?
From a quick glance, looks like https://github.com/rust-lang/rustfmt/pull/5290 should be merged and https://github.com/rust-lang/rustfmt/issues/5269 fixed before that.
Is there anything else?

**Comment by calebcartwright** [comment]
> Hey! I'm happy to see the rustfmt project active once again.

I assume this was said with the best of intent and while I appreciate the positive sentiment, I'd dispute the characterization. However, that's not a topic I'm interested in delving into, not to mention that doing so here would be off topic anyway

> From a quick glance, looks like https://github.com/rust-lang/rustfmt/pull/5290 should be merged and https://github.com/rust-lang/rustfmt/issues/5269 fixed before that.
Is there anything else?

I won't say definitively that's the ultimate, exhaustive list of blocking issues, but yes, it's accurate that those remain some of the known bodies of work pending (and blocking) for stabilization of this option.

That's been the case for a while, and at least in the case of #5269, is unlikely to change any time soon unless someone else works on them; neither Yacin nor I are likely to be able to allocate any of our rustfmt-bandwidth on those for the foreseeable future.

**Comment by popzxc** [comment]
Sorry that my remark turned out to be inaccurate, I was just happy to see that PRs are being actively merged & discussed. I hope it didn't appear insensitive.

Then, I guess, the main question I have is: do you see any blockers here besides development effort (that can be done by external contributors)? E.g. is there enough capacity to discuss topics, accept/reject proposed solutions, etc (basically everything *but* submitting PRs)?

I'm interested in working on it, but certainly don't want to overburden the rustfmt maintainers.

**Comment by 9999years** [comment]
I've added a list of open issues regarding `imports_granularity` to the original ticket. Please let me know if there's others to add to the list.

**Comment by yanchith** [comment]
Hi all. I hope I am not too late with the feedback. I would like to use `imports_granularity = "Module"`, as in theory it is exactly what I'd like to see.

However, I have a couple of issues with how it works as of nightly-2024-05-13.

1) I think the following diff shows bugs, or at least really unexpected features.

```rust
use core::mem::{self, MaybeUninit};
```
... becomes ...

```rust
use core::mem::{  // HorizontalVertical is not respected
    MaybeUninit,
   {self},        // self became wrapped in additional {} 
};
```

2) This is technically not a bug, but I'd like to have module imports as separate use statements and only merge non-module items of the last depth level. Maybe this is another imports_granularity? "ModulesWithModuleUsesAsOne". 

```rust
use core::fmt;
use core::slice;
```

... becomes ...

```rust
use core::{fmt, slice};
```

**Comment by mnkhouri** [comment]
@yanchith I filed https://github.com/rust-lang/rustfmt/issues/6191 so that your point 3. can be discussed, since I also expected the same result you described

**Comment by Blub** [comment]
Is there any reason not to stabilize *parts* of this? Eg. we'd like to use `Item` as it seems quite straightforward and the most merge-conflict-friendly option of them all. (Anything that re-flows entire blocks of imports is a PITA for `git merge` & manual reviews)

**Comment by pront** [comment]
We would love to use this option for the https://github.com/vectordotdev/vector codebase.

**Comment by 9999years** [comment]
Do we still need #5296? The PRs it backports, #4406 and #4536, were merged 5 years ago. Surely we don't need backports for them?

**Comment by 9999years** [comment]
I've added some more issue links in the issue description but none of them apply to `imports_granularity = "Item"`.

**Comment by SimonSapin** [comment]
It there precedent for stabilizing only some values of an option?

**Comment by Diggsey** [comment]
Seems like the issues relating to the `group_imports` option should not block stabilizing this, since that option is also unstable?

**Comment by sdevrs** [comment]
Hi @Manishearth, I'm pinging you for an update since you're listed as the [Team Leader](https://rust-lang.org/governance/teams/dev-tools/).

I've been using `imports_granularity = "Crate"` for years and it's been working great! Many others have also shared similar sentiments.

It would be great if the Dev Tools Team could review the requirements for getting this stabilized and increase the priority for doing so as it's immensely useful and impactful.

I think formatting and compacting the imports is an expected feature of the formatter, and its absence is being felt widely and has been for a long time.

I stumbled upon this feedback this morning:
"This is just ANNOYING...This is the kind of thing that makes future conflicts harder for me to deal with." - Linus Torvalds

You can read more about Linus' experience with it [here](https://lore.kernel.org/lkml/CAHk-=wgO7S_FZUSBbngG5vtejWOpzDfTTBkVvP3_yjJmFddbzA@mail.gmail.com/T/#me533a148abe97c29e0e7150508c42345b2a64e13).

It feels like we're waiting on perfection instead of iterating towards it, but I know you're being cautious as you do the latter, while balancing all your other work.

Thanks for all the work that you and the team do!

Looking forward to your update and hopefully we’ll see `imports_granularity` in the [release notes](https://doc.rust-lang.org/releases.html) soon.

**Comment by Manishearth** [comment]
Note: While I do lead the devtools team, devtools subteams are largely autonomous. The rustfmt team is [a rather small team ](https://rust-lang.org/governance/teams/dev-tools/#team-rustfmt) led by @calebcartwright . My comment here is my own opinion, not that of the rustfmt team, whose ultimate decision matters here.

So, I want this too. Making it easier to avoid conflicts is great; and in general the whole set of import management formatting settings would be amazing to see stabilized.

But, looking at this issue it's pretty clear why this feature hasn't merged yet: this feature has some serious bugs. In some cases it changes the meaning of the code, and there's non-idempotency in other cases (some, but not all, requiring a second unstable feature). This isn't "waiting on perfection", this is about table stakes guarantees that `rustfmt` wishes to provide, that it currently provides rather successfully. I find it hard to see `rustfmt` compromising on that.

This is a small team of volunteers: if people want to move this along, I recommend looking in to what it takes to fix these remaining issues. Specifically, #6028 needs to be iterated on further, and #6243 needs to be fixed. The others deal with interactions with a different unstable feature and it may be possible to decide that those bugs are the responsibility of the other feature to fix, I don't know.


As for Linus' comment ... Ultimately I don't personally feel the need to give much weight to a tantrum thrown by Prominent Open Source Man Known To Throw Tantrums. Like yeah, I already know this is a pain point, Linus being annoyed about it doesn't really tell me anything.


@ojeda has been doing a great job of working with the Rust project to help it understand Linux's needs around feature work and stabilization. I'm curious to hear what he thinks of this issue; perhaps he can help find people to work on it if he considers this a priority for Linux.

**Comment by eveeifyeve** [comment]
What can I do to help move this along, because I want to contribute. 
But don't know what I can do to help with this?

Like what are we waiting on or is everything implemented?

**Comment by Diggsey** [comment]
@Eveeifyeve quoting the comment above...

> This is a small team of volunteers: if people want to move this along, I recommend looking in to what it takes to fix these remaining issues. Specifically, https://github.com/rust-lang/rustfmt/pull/6028 needs to be iterated on further, and https://github.com/rust-lang/rustfmt/issues/6243 needs to be fixed. The others deal with interactions with a different unstable feature and it may be possible to decide that those bugs are the responsibility of the other feature to fix, I don't know.

**Comment by eveeifyeve** [comment]
> [@Eveeifyeve](https://github.com/Eveeifyeve) quoting the comment above...
> 
> > This is a small team of volunteers: if people want to move this along, I recommend looking in to what it takes to fix these remaining issues. Specifically, [#6028](https://github.com/rust-lang/rustfmt/pull/6028) needs to be iterated on further, and [#6243](https://github.com/rust-lang/rustfmt/issues/6243) needs to be fixed. The others deal with interactions with a different unstable feature and it may be possible to decide that those bugs are the responsibility of the other feature to fix, I don't know.

Thankyou for quoting this, however it's more reviewing code of https://github.com/rust-lang/rustfmt/pull/6677 and https://github.com/rust-lang/rustfmt/pull/6028 and not actually implementations that are needed and haven't been done in a pr yet. I just want to help clear off the back log for this issue. 

So I assume everything is implemented right?

**Comment by David-OConnor** [comment]
Would love to have this feature in stable! Have been using it for years. _"Bugs are bad, but perfect is the enemy of good enough."_ - RDML Grace Hopper

**Comment by pront** [comment]
FWIW I used this option on a large codebase and it didn't cause any bugs: https://github.com/vectordotdev/vector/pull/23730

**Comment by Manishearth** [comment]
@Eveeifyeve 

> Thankyou for quoting this, however it's more reviewing code of [#6677](https://github.com/rust-lang/rustfmt/pull/6677) and [#6028](https://github.com/rust-lang/rustfmt/pull/6028) and not actually implementations that are needed and haven't been done in a pr yet. I just want to help clear off the back log for this issue.

No, if you look at #6028 it's stalled on further work from the PR author. Someone else can pick that up if they want.

#6677 is indeed a new PR that attempts to fix #6243, it will likely take some time to be reviewed by team members. At a glance, it looks like the right approach!

Ideally, help on the other items in the checklist at the top of this issue would also be good. Most of the otherspertain to interactions with `group_imports`, could in theory be punted to the stabilization of _that_ feature, but we're not yet at the point where that decision needs to be made. It would be nice if it didn't have to be.

------------------


@David-OConnor This isn't a matter of "perfect is the enemy of the good". idempotency and _not breaking code_ are table stakes for rustfmt. There are many cases where a feature request is open for many years due to trying for perfection, I don't think this is one of them.


----------

Also, since this issue is gaining visibility, for people coming to this issue now: please do not post "me too!" comments. I've explained what needs to be done to bring this over the finish line, people saying they want this does not move the needle.


In general when it comes to issues like this when there's been a period of inactivity, asking what is needed to take it over the finish line is fine, just saying you want a thing is noise that everyone will just ignore in their notifications.

**Comment by ojeda** [comment]
Thanks Manish for the kind words and for commenting on the issue.

I have been thinking about this, and I ended up proposing the trailing comment workaround for the time being, and pinged the `rustfmt` maintainers to see what can be done for the future.

As for contributing: I typically mention that one can contribute to Rust for Linux in many ways, not just on the kernel side -- this is one more way to do so, so if someone wants to help here, please ping me!

**Comment by Manishearth** [comment]
Thank you! It sounds like you do have interest in this feature, but you have a workaround for now.

We're working on improved maintenance of rustfmt and may be able to at least merge the bugfix PRs soon. We might let them bake for a while before deciding on unflagging.

> As for contributing: I typically mention that one can contribute to Rust for Linux in many ways, not just on the kernel side -- this is one more way to do so, so if someone wants to help here, please ping me!



Very, very appreciated! 😄

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rustfmt/issues/6717 seems like a major issue in `imports_granularity = One` that would require it to understand the import tree to be properly fixed (which is beyond the scope of formatting). That's worrying. There might be ways around it.

**Comment by Max-V-Seltz** [comment]
Hello everybody,

I have been following this feature since 2022 and would like to help move it toward stabilization.

As a first-time contributor to rustfmt, are there any specific blocking issues that would be a good starting point?
I’m happy to take on a smaller task to get familiar with the codebase before tackling more complex stabilization work.

**Comment by David-OConnor** [comment]
Let's do it! This would be a net win. Group Imports too. 

```toml
imports_granularity = "Crate"
group_imports = "StdExternalCrate"
```

**Comment by shssoichiro** [comment]
I see there are also some open PRs such as https://github.com/rust-lang/rustfmt/pull/6723 which have not been reviewed since November. Is there something we can do to aid in triaging these as well? I'd very much like to see `imports_granularity` and `group_imports` both stabilized.

**Comment by Kobzol** [comment]
The maintenance situation of rustfmt has been complicated for some time, see discussion on Zulip (e.g. https://rust-lang.zulipchat.com/#narrow/channel/357797-t-rustfmt/topic/Best.20way.20to.20contribute/near/544415538). I think there is some progress ongoing to restart development a bit.

**Comment by pfsavm** [comment]
What does it need to finally bring this to `stable`?

We would like to use this in production and nightly is always a hurdle.

**Comment by xStrom** [comment]
> What does it need to finally bring this to `stable`?

Check the first post of this tracking issue for a list of open issues.

**Comment by pfsavm** [comment]
> Check the first post of this tracking issue for a list of open issues.

Thanks, I see the issues, but I'm still not clear about the way forward.
I want to help progress this but there are a lot of open questions like:
- Is every not-fully-functional edge-case, that someone found, really a blocker? Or wouldn't it be more helpful to most people if it is stablized regardless and minor edge cases can be fixed afterwards, if relevant at all (and not purely constructed, highly theoretical)
- Some issues are lying around for years, what's the action that I can do today to progress this?

I think a lot of people share the opinion that these features are good enough and the missing stablization blocks a wider adoption especially in company-production environments...

**Comment by Manishearth** [comment]
> Is every not-fully-functional edge-case, that someone found, really a blocker?

Yes. See my comment [here](https://github.com/rust-lang/rustfmt/issues/5083#issuecomment-3929295921) about the stability guarantees rustfmt is under. We cannot just stabilize and "fix issues later". This makes stabilization a pretty high bar.

We _can_ determine that some of these issues are not worth _ever_ fixing (which is fine from the stability POV). This is a potential outcome of many of these sub discussions, but that outcome must happen through discussion and design work; exploring the design space.

For a lot of these issues so far nobody has really explored the design space like that.

A fair number of these issues also affect rustfmt's core property that you should always be able to run it without the behavior/correctness of your code changing. A lot of these sub issues are pretty major, from that angle.

The rustfmt team is small and made up of volunteers. "I want this" and "why isn't this done yet" (when there is a clear issue) are not productive comments to make on our tracking issues: we waste time engaging with them, and they clutter up the threads.

More people wanting this does not magically make the work happen.

The rustfmt team is also focusing on fixing the syncs with rustc which is a far more pressing issue than new features.

------


> Some issues are lying around for years, what's the action that I can do today to progress this?

Some issues are straight up bugs and needs someone to go through the code and try to fix them. I think https://github.com/rust-lang/rustfmt/issues/6027 and https://github.com/rust-lang/rustfmt/issues/6195 are in that bucket, though the first one has an open PR. Might be able to ask in Zulip if someone feels comfortable reviewing it.


Some issues need some design work. For example, https://github.com/rust-lang/rustfmt/issues/6717 is a pretty fundamental issue with the design of this formatting mode, and people need to explore the design space of options and see which one the rustfmt team is happy with. https://github.com/rust-lang/rustfmt/issues/6241 is similar.

**Comment by David-OConnor** [comment]
RustFmt bros: Thank you for maintaining the high standard; the quality of built-in tooling is one of Rust's strongest assets; it is unmatched.

To those like these formatting features but don't like typing "+nightly": Maybe script it or something, or set up a shortcut in the IDE? I have been using this for years as it's a great feature I haven't hit errors with.  

I think having "nightly" available is a good compromise to make things workable for both perspectives.

**Comment by shssoichiro** [comment]
This is slightly off-topic, but I'll explain the issue with using it behind nightly: For those who want to use it on public projects, it presents confusion to new contributors. Every new contributor has to be told to add `+nightly` when they run `cargo fmt`.

That being said, it still seems like the correct approach. It's better than stabilizing something that is broken and then never being able to change it ever.

**Comment by Manishearth** [comment]
The vote of support is appreciated, but I'm going to aggressively hide comments that are not actively working towards resolving this issue: this thread is long and hard to follow already. I also cleaned up some of the past thread before I got bored. Might do more later.

"I want this too" and "why isn't this done yet" (when there are open issues) are not constructive comments. I'm leaving my previous comment unhidden because it provides _some_ direction to people interested in contributing. Joining the Zulip channel for rustfmt and talking about individual issues is encouraged.

**Comment by willbuckner** [comment]
Not to be that guy, but this issue is 5 years old now. Is there anything I can do to help? Formatting with nightly is really not ideal

**Comment by Manishearth** [comment]
Since this got collapsed into the "show more" by github, please see my comment above for a status and ways people can help

https://github.com/rust-lang/rustfmt/issues/4991#issuecomment-3369792412

Basically:

 - There are open issues that need work. Some need design discussion.
 - There are open PRs for some of the issues, however the rustfmt team is heavily backlogged on review, and reviews for unstable features are low priority. Community reviews of such PRs are appreciated, but ideally you can get involved on zulip and help the team with other things too.

In general until the rustfmt team is healthy enough to sustain _regular_ maintainance work it is unlikely that we shall see progress on unstable features.
